**Question 1 [ 40 Points ] - Support Vector Machines (SVMs)** 

[25 points ] Download this dataset, split it as a 80% training and 20% test set. and implement the support vector algorithm from scratch using Numpy and Pandas.

[10 points ] Report the accuracies for the train and test sets. Comment on whether your model has overfit.

[5 points] Test your model performance with the scikit-learn model. Comment on the difference in accuracy. 


In [1]:
! pip install numpy
! pip install sklearn
! pip install pandas

Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable


In [2]:
import numpy as np
import pandas as pd

In [3]:
columns = ["sepal length", "sepal width", "petal length", "petal width", "label"]
df = pd.read_csv('./iris.data', names=columns)

converted = []
for i in range(len(df['label'])):
    if df['label'][i] == "Iris-setosa":
        converted.append(0)
    elif df['label'][i] == "Iris-versicolor":
        converted.append(1)
    else:
        converted.append(2)
df['converted'] = converted
print(df)

     sepal length  sepal width  petal length  petal width           label  \
0             5.1          3.5           1.4          0.2     Iris-setosa   
1             4.9          3.0           1.4          0.2     Iris-setosa   
2             4.7          3.2           1.3          0.2     Iris-setosa   
3             4.6          3.1           1.5          0.2     Iris-setosa   
4             5.0          3.6           1.4          0.2     Iris-setosa   
..            ...          ...           ...          ...             ...   
145           6.7          3.0           5.2          2.3  Iris-virginica   
146           6.3          2.5           5.0          1.9  Iris-virginica   
147           6.5          3.0           5.2          2.0  Iris-virginica   
148           6.2          3.4           5.4          2.3  Iris-virginica   
149           5.9          3.0           5.1          1.8  Iris-virginica   

     converted  
0            0  
1            0  
2            0  
3      

In [4]:
class SVM:

    def __init__(self, number_of_iterations=1000, lamb=0.001, alpha=0.01, c = 1, main_class=0):
        self.weights = None
        self.bias = None
        self.alpha = alpha
        self.number_of_iterations = number_of_iterations
        self.lambda_param = lamb
        self.c = c
        self.main_class = main_class

    def convert(self, y):
        y_new = [0] * len(y)
        for i in range(len(y)):
            if y[i] != self.main_class:
                y_new[i] = -1
            elif y[i] == self.main_class:
                y_new[i] = 1
        
        return y_new

    def fit(self, X, y):
        n_samples, n_features = X.shape
        self.weights = np.zeros(n_features)
        self.bias = 0

        converted_y = self.convert(y)


        for _ in range(self.number_of_iterations):
            for i in range(len(X)):
                condition = converted_y[i] * \
                    (np.dot(X[i], self.weights) - self.bias) >= self.c
                if condition:
                    self.weights -= self.alpha * \
                        (2 * self.lambda_param * self.weights)
                else:
                    self.weights -= self.alpha * \
                        (2 * self.lambda_param *
                         self.weights - np.dot(X[i], converted_y[i]))
                    self.bias -= self.alpha * converted_y[i]

    def predict(self, X):
        ans = []
        estimate = np.dot(X, self.weights) - self.bias
        for i in estimate:
            if np.sign(i)<0:
                if self.main_class == 0:
                    ans.append(1)
                elif self.main_class == 1:
                    ans.append(2)
                elif self.main_class == 2:
                    ans.append(0)
            else: 
                ans.append(self.main_class)
        
        return ans


In [5]:
df.head()

sepal length  sepal width  petal length  petal width        label  \
0           5.1          3.5           1.4          0.2  Iris-setosa   
1           4.9          3.0           1.4          0.2  Iris-setosa   
2           4.7          3.2           1.3          0.2  Iris-setosa   
3           4.6          3.1           1.5          0.2  Iris-setosa   
4           5.0          3.6           1.4          0.2  Iris-setosa   

   converted  
0          0  
1          0  
2          0  
3          0  
4          0

In [6]:
from sklearn.model_selection import train_test_split
X = np.array(df.iloc[:,0:4])
y = np.array(df['converted'])
print(X, y)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=101)

[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.  3.2 1.2 0.2]
 [5.5 3.5 1.3 0.2]
 [4.9 3.1 1.5 0.1]
 [4.4 3.  1.3 0.2]
 [5.1 3.4 1.5 0.2]
 [5.  3.5 1.3 0.3]
 [4.5 2.3 1.3 0.3]
 [4.4 3.2 1.3 0.2]
 [5.  3.5 1.6 0.6]
 [5.1 3.8 1.9 0.4]
 [4.8 3.  1.4 0.3]
 [5.1 3.8 1.6 0.2]
 [4.6 3.2 1.4 0.2]
 [5.3 3.7 1.5 0.2]
 [5.  3.3 1.4 0.2]
 [7.  3.2 4.7 1.4]
 [6.4 3.2 4.5 1.5]
 [6.9 3.1 4.

In [7]:
# Model iris-setosa => 0
model = SVM(number_of_iterations=1000, lamb=0.001, alpha=0.01, c = 1, main_class=0)
model.fit(X_train, y_train)
y_pred_0 = model.predict(X_test)
print(y_pred_0)

[0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0]


In [8]:
# Model iris-versicolor => 1
model = SVM(number_of_iterations=1000, lamb=0.001, alpha=0.01, c = 1, main_class=1)
model.fit(X_train, y_train)
y_pred_1 = model.predict(X_test)
print(y_pred_1)

[2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 1, 1, 1, 1, 1, 2, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1]


In [9]:
# Model iris-virginica => 2
model = SVM(number_of_iterations=1000, lamb=0.001, alpha=0.01, c = 1, main_class=2)
model.fit(X_train, y_train)
y_pred_2 = model.predict(X_test)
print(y_pred_2)

[0, 0, 0, 0, 0, 2, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0]


In [10]:
final_predictions = []
for i in range(len(y_pred_0)):
    map = {0:0, 1:0, 2:0}
    map[y_pred_0[i]] += 1
    map[y_pred_1[i]] += 1
    map[y_pred_2[i]] += 1

    m = max(map.values())
    if map[0] == m:
        final_predictions.append(0)
    elif map[1] == m:
        final_predictions.append(1)
    else:
        final_predictions.append(2)
    

print(final_predictions)

[0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0]


In [11]:
# Model iris-setosa => 0
model = SVM(number_of_iterations=1000, lamb=0.001, alpha=0.01, c = 1, main_class=0)
model.fit(X_train, y_train)
y_pred_0 = model.predict(X_train)
print(y_pred_0)

[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1]


In [12]:
# Model iris-versicolor => 1
model = SVM(number_of_iterations=1000, lamb=0.001, alpha=0.01, c = 1, main_class=1)
model.fit(X_train, y_train)
y_pred_1 = model.predict(X_train)
print(y_pred_1)

[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 2, 1, 1, 1, 1, 2, 1, 2, 1, 1, 2, 1, 2, 1, 1, 1, 1, 1, 2, 1, 1, 2, 1, 1, 1, 2, 2, 2, 1, 2, 2, 1, 1, 1, 2, 1, 2, 1, 2, 1, 1, 2, 1, 1, 1, 2, 1, 2, 1, 1, 1, 1, 1, 1, 1, 2, 1, 2, 1, 1, 2, 1, 1, 1, 1, 1, 2, 2, 1, 1, 1, 2, 2, 1, 1, 2, 1, 1, 2, 1, 1, 2, 2, 2, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 2, 1]


In [13]:
# Model iris-virginica => 2
model = SVM(number_of_iterations=1000, lamb=0.001, alpha=0.01, c = 1, main_class=2)
model.fit(X_train, y_train)
y_pred_2 = model.predict(X_train)
print(y_pred_2)

[2, 0, 2, 0, 2, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 2, 0, 0, 2, 0, 0, 0, 0, 0, 0, 2, 2, 0, 0, 0, 0, 2, 2, 0, 2, 0, 2, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 2, 2, 2, 0, 0, 2, 0, 0, 2, 2, 0, 0, 0, 2, 2, 0, 2, 0, 0, 2, 0, 2, 0, 0, 0, 0, 0, 0, 2, 0, 2, 0, 2, 2, 2, 0, 0, 0, 2, 0, 0, 0, 2, 2, 2, 2, 0, 2, 0, 2, 0, 2, 2, 0, 2, 0, 0, 0, 0, 0, 0, 0]


In [14]:
final_predictions2 = []
for i in range(len(y_pred_0)):
    map = {0:0, 1:0, 2:0}
    map[y_pred_0[i]] += 1
    map[y_pred_1[i]] += 1
    map[y_pred_2[i]] += 1

    m = max(map.values())
    if map[0] == m:
        final_predictions2.append(0)
    elif map[1] == m:
        final_predictions2.append(1)
    else:
        final_predictions2.append(2)
    

print(final_predictions2)

[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 2, 1, 0, 1, 0, 1, 1, 1, 2, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 2, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 2, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 2, 1, 1, 0, 2, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1]


In [15]:
from sklearn.metrics import accuracy_score

print("My SVM Train data accuracy is :", accuracy_score(final_predictions2, y_train))
print("My SVM  Test data accuracy is :", accuracy_score(final_predictions, y_test))
print("The model is not overfitted")

My SVM Train data accuracy is : 0.7
My SVM  Test data accuracy is : 0.7333333333333333
The model is not overfitted


In [16]:
from sklearn import svm

model = svm.SVC(C=1.0)
model.fit(X_train, y_train)
test_predictions = model.predict(X_test)
train_predictions = model.predict(X_train)

print("SKLearn SVM Test data accuracy is :", accuracy_score(test_predictions, y_test))
print("SKLearn SVM  Train data accuracy is :", accuracy_score(train_predictions, y_train))

"""
The difference in the accuracy could be because of the OVR (one-versus-rest) approach
taken by me and Sklearn internal classifier handled multi class data differently
It is also possible that the learning rate and number of iterations might be different
due to which the accuracy is different
"""

SKLearn SVM Test data accuracy is : 0.9666666666666667
SKLearn SVM  Train data accuracy is : 0.95


'\nThe difference in the accuracy could be because of the OVR (one-versus-rest) approach\ntaken by me and Sklearn internal classifier handled multi class data differently\nIt is also possible that the learning rate and number of iterations might be different\ndue to which the accuracy is different\n'

Training accuracy : 0.7

Test accuracy : 0.733

**Question 2 [ 40 Points ] - Decision Trees** 

a. [5 points] Complete the test_split function.

b. [5 points] Complete the gini_index function.

c. [5 points] Complete the get_split function.

d. [15 points] Complete the split function.

e. [10 points] Print the tree. 

In [17]:
from sklearn.datasets import load_iris
import numpy as np

In [18]:
def test_split(index, value, dataset):
    left, right = list(), list()
    for row in dataset:
        if row[index] < value:
            left.append(row)
        else:
            right.append(row)
    return left, right


In [19]:
def gini_index(groups, classes):
    # count all samples at split point
    n_instances = float(sum([len(group) for group in groups]))
    # sum weighted Gini index for each group
    gini = 0.0
    
    for group in groups:
        try:
            count = 0

            for class_label in classes:
                class_label_count = 0
            
                for i in group:
                    if i[-1] == class_label:
                        class_label_count += 1
                count += (class_label_count / len(group)) ** 2

            gini += (1 - count) * (len(group) / n_instances)
        except ZeroDivisionError:
            continue
    return gini


In [20]:
# Select the best split point for a dataset
def get_split(dataset):
    class_values = list(set(row[-1] for row in dataset))
    b_index, b_value, b_score, b_groups = 999, 999, 999, None
    
    for index in range(len(dataset[0])-1):
        for row in dataset:
            # Hint : Call the test_split function here.
            groups = test_split(index, row[index], dataset)
            # Hint : Call the gini_index function here.
            gini = gini_index(groups, class_values)
            if gini < b_score:
                b_index, b_value, b_score, b_groups = index, row[index], gini, groups
    return {'index': b_index, 'value': b_value, 'groups': b_groups}


In [21]:
# Create a terminal node value
def to_terminal(group):
	outcomes = [row[-1] for row in group]
	return max(set(outcomes), key=outcomes.count)

In [22]:
# Create child splits for a node or make terminal
#Hint : Just call the to_terminal and get_split functions defined above. 

def split(node, max_depth, min_size, depth):
    left, right = node['groups']
    del (node['groups'])

    # check for a no split
    if not left or not right:
        node['left'] = node['right'] = to_terminal(left + right)
        return

    # check for max depth
    if depth >= max_depth:
        node['left'], node['right'] = to_terminal(left), to_terminal(right)
        return

    # process left child
    if len(left) <= min_size:
        node['left'] = to_terminal(left)
    else:
        node['left'] = get_split(left)
        split(node['left'], max_depth, min_size, depth+1)
    
    # process right child
    if len(right) <= min_size:
        node['right'] = to_terminal(right)
    else:
        node['right'] = get_split(right)
        split(node['right'], max_depth, min_size, depth+1)

In [23]:
def build_tree(train, max_depth, min_size):
    root = get_split(train)
    split(root, max_depth, min_size, 1)
    return root

In [24]:
def print_tree(node, depth=0):
    if isinstance(node, dict):
        print('%s[X%d < %.3f]' %
              ((depth*' ', (node['index']+1), node['value'])))
        print_tree(node['left'], depth+1)
        print_tree(node['right'], depth+1)
    else:
        print('%s[%s]' % ((depth*' ', node)))

In [25]:
iris= load_iris()

X= np.array(iris.data)
y= np.array(iris.target).reshape(-1,1)

data = np.append(X,y,axis=1)

In [26]:
tree = build_tree(data, 1, 1)
print_tree(tree)

[X3 < 3.000]
 [0.0]
 [1.0]


**Question 3 [ 20 Points ] - Random Forests and Boosting**


In [27]:
col_names = ['buying', 'maint', 'doors', 'persons', 'lug_boot', 'safety', 'class']

df = pd.read_csv('./car.data', names=col_names)

X = df.drop(['class'], axis=1)
y = df['class']

In [28]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.33, random_state = 42)

In [29]:
! pip install category-encoders

Defaulting to user installation because normal site-packages is not writeable


In [30]:
import category_encoders

# We need to convert the strings into numbers
encoder = category_encoders.OrdinalEncoder(cols = col_names[:len(col_names) - 1])
X_train, X_test = encoder.fit_transform(X_train), encoder.transform(X_test)

In [31]:
from sklearn.ensemble import RandomForestClassifier

classifier = RandomForestClassifier(random_state=0)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)
y_pred_train = classifier.predict(X_train)

In [32]:
print('Train Accuracy is :', accuracy_score(y_train, y_pred_train))
print('Test Accuracy is :', accuracy_score(y_test, y_pred))

Train Accuracy is : 1.0
Test Accuracy is : 0.9457092819614711


In [33]:
from sklearn.ensemble import GradientBoostingClassifier

gradient_boost_classifier = GradientBoostingClassifier()
gradient_boost_classifier.fit(X_train, y_train)
y_pred = gradient_boost_classifier.predict(X_test)
y_pred_train = gradient_boost_classifier.predict(X_train)

In [34]:
print('Train Accuracy is :', accuracy_score(y_train, y_pred_train))
print('Test Accuracy is :', accuracy_score(y_test, y_pred))

Train Accuracy is : 0.9913569576490925
Test Accuracy is : 0.968476357267951
